Add a backoff mechanism for the CommunicationTokenCredential (#25031)
Azure/azure-sdk-for-net: sdk/communication/Azure.Communication.Common/README.md

Optionally, you can enable proactive token refreshing where a fresh token will be acquired as soon as the
previous token approaches expiry. Using this method, your requests are less likely to be blocked to acquire a fresh token:

> Optionally, you can provide a time span before token expiry that `tokenRefresher` will be called if `refreshProactively` is true. For example, setting it to 5min means that 5min before the cached token expires, the proactive refresh will request a new token.

Reviewer: ```suggestion
Optionally, you can adjust the time span before token expiry that will trigger the proactive refreshing of the token. For example, if the proactive refreshing is enabled ('refreshProactively' is true), setting the time to a value equal to 5 minutes means that 5 minutes before the cached token expires, the proactive refresh will request a new token by calling the 'tokenRefresher' callback.
```

Reviewer: @AikoBB have you intentionally not committed this one?

Author: guess I missed this one

> The default value is 10min.

Reviewer: ```suggestion
The default value is 10 minutes.
```

```C# Snippet:CommunicationTokenCredential_CreateRefreshableWithoutInitialToken
using var tokenCredential = new CommunicationTokenCredential(
    new CommunicationTokenRefreshOptions(
        refreshProactively: true, // Indicates if the token should be proactively refreshed in the background or only on-demand

>         refreshTimeBeforeTokenExpiry: new TimeSpan(0, 0, 5, 0),

Reviewer: ```suggestion
        refreshTimeBeforeTokenExpiry: new TimeSpan(0, 0, 5, 0), // Proactive refreshing will be triggered 5 minutes before the token expires
```

>         refreshTimeBeforeTokenExpiry: new TimeSpan(0, 0, 5, 0), // Proactive refreshing will be triggered 5 minutes before the token expires

Reviewer 2: Have we considered making `refreshTimeBeforeTokenExpiry` a public property instead of a constructor argument similar to `AsyncTokenRefresher` and `InitialToken`? It's similar optional. I think it should be a property because usually a C# options bag only has the default constructor and properties which are all considered optional. But here we had to introduce a constructor because `tokenRefresher` is required.

I suggest to remove the new constructor and make the property public with a setter. Then the Contoso code would look like

```c#
using var tokenCredential = new CommunicationTokenCredential(
    new CommunicationTokenRefreshOptions(
        refreshProactively: true, // Indicates if the token should be proactively refreshed in the background or only on-demand
        tokenRefresher: cancellationToken => FetchTokenForUserFromMyServer("bob@contoso.com", cancellationToken))
    {
        RefreshTimeBeforeTokenExpiry = new TimeSpan(0, 0, 5, 0), // Proactive refreshing will be triggered 5 minutes before the token expires
        AsyncTokenRefresher = cancellationToken => FetchTokenForUserFromMyServerAsync("bob@contoso.com", cancellationToken)
    });
```

Sorry for not seeing this earlier.

Reviewer 2: Here's the applicable guideline: https://azure.github.io/azure-sdk/dotnet_introduction.html#dotnet-parameters

> The Options class is designed similarly to .NET custom attributes, where required service method parameters are modeled as Options class constructor parameters and get-only properties, and optional parameters are get-set properties.

Author: @DominikMe thank you so much for your comments and a huge impact on the improvements😊 Your comment is fixed for .net, should we change it as well for other languages?

Reviewer: @DominikMe makes sense, thanks :)
@AikoBB if it's applicable to Java (check the guidelines) then yes. JS/TS works differently and Python's not supposed to have option bags.
